### PR TITLE
Subgraph refactor

### PR DESCRIPTION
**Pull request by aj-prime**
Refactored sub-graph parallelism code
Removed Old sub-grid creation code
Added unit tests for cross_grid layers
Removed old sub-graph parallelism variables
Modified sub-graph parallelism example code for Transformers

### CODE REVIEW

File: ci_test/unit_tests/test_unit_subgrid_cross_grid_sum.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def setup_experiment(lbann, weekly):
 
     """
     
-    pytest.skip('Skip - Temporarily skipping subgraph tests')
+    #pytest.skip('Skip - Temporarily skipping subgraph tests')
```
Comment (tbennun) [inline ci_test/unit_tests/test_unit_subgrid_cross_grid_sum.py:47]:
```suggestion
```

File: ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+import tools
+
+# ==============================================
+# Objects for Python data reader
+# ==============================================
+# Note: The Python data reader imports this file as a module and calls
+# the functions below to ingest data.
+
+# Data
+np.random.seed(202201184)
```
Comment (tbennun) [inline ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py:21]: Update date

File: ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_slice.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import tools
+
+# ==============================================
+# Objects for Python data reader
+# ==============================================
+# Note: The Python data reader imports this file as a module and calls
+# the functions below to ingest data.
+
+# Data
+np.random.seed(202201184)
```
Comment (tbennun) [inline ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_slice.py:21]: Update date

File: ci_test/unit_tests/test_unit_subgrid_split.py [hunk 4]
```diff
@@ -86,19 +84,19 @@ def construct_model(lbann):
...
     y1 = lbann.Sin(x1)
     y2 = lbann.Cos(x2)
     y = lbann.Sum(
         lbann.Identity(y1),
         lbann.Identity(y2),
-        parallel_strategy = {'grid_tag':0, 'enable_subgraph':True, 'sub_branch_tag':0})
+        parallel_strategy = {'grid_tag':0})
     z = lbann.L2Norm2(y)
     obj.append(z)
     metrics.append(lbann.Metric(z, name='obj'))
```
Comment (tbennun) [inline ci_test/unit_tests/test_unit_subgrid_split.py:102]: no need to set subgraph communication in this test?

File: ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
+import functools
```
Comment (tbennun) [inline ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py:1]: General note: should the test file names change to subgraph?

File: include/lbann/layers/layer.hpp [hunk 6]
```diff
@@ -866,9 +870,11 @@ class Layer
   /// @todo Remove
 
   SubGraphCommunication subgraph_communication_method = PT2PT;
-
+  //model-level sub-graph parallelism
```
Comment (tbennun) [inline include/lbann/layers/layer.hpp:873]:
```suggestion
  // Model-level sub-graph parallelism
```

File: include/lbann/layers/layer.hpp [hunk 7]
```diff
@@ -866,9 +870,11 @@ class Layer
...
   bool apply_subgraph_parallelism = false;
-
+  //Layer-level sub-graph execution
```
Comment (tbennun) [inline include/lbann/layers/layer.hpp:875]:
```suggestion
  // Layer-level sub-graph execution
```

File: include/lbann/layers/layer.hpp [hunk 8]
```diff
@@ -866,9 +870,11 @@ class Layer
...
+  bool m_subgraph_parallelism_execution = false;
+  //Process-level sub-graph execution
```
Comment (tbennun) [inline include/lbann/layers/layer.hpp:877]:
```suggestion
  // Process-level sub-graph execution
```

Comment (tbennun) [inline include/lbann/layers/layer.hpp:877]: What does "Process-level" mean in this context?

File: include/lbann/layers/transform/cross_grid_sum.hpp [hunk 9]
```diff
@@ -158,8 +172,43 @@ class cross_grid_sum_layer final : public data_type_layer<TensorDataType>
   void bp_setup_gradient_wrt_inputs(El::Int mini_batch_size) final
   {
     int rank = El::mpi::Rank(this->get_subgrid_comm());
-    const auto& gradient_wrt_output = this->get_prev_error_signals(rank);
-    auto& gradient_wrt_input = this->get_error_signals(rank);
+    auto parents = this->get_parent_layers();
+    auto childs = this->get_child_layers();
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum.hpp:176]:
```suggestion
    auto children = this->get_child_layers();
```

File: include/lbann/layers/transform/cross_grid_sum.hpp [hunk 10]
```diff
@@ -158,8 +172,43 @@ class cross_grid_sum_layer final : public data_type_layer<TensorDataType>
...
+    }
+    tag = tag_parent-1;
+    for(int i=0; i<childs.size();i++)
+    {
+      if (this->get_prev_error_signals(i).Grid().InGrid())
+        tag_child = childs[i]->get_grid_tag();
+    }
+
+    // int tag_parent = parents[rank]->get_grid_tag();
+    // int tag_child = childs[rank]->get_grid_tag();
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum.hpp:196]:
```suggestion
```

File: include/lbann/layers/transform/cross_grid_sum.hpp [hunk 11]
```diff
@@ -158,8 +172,43 @@ class cross_grid_sum_layer final : public data_type_layer<TensorDataType>
...
+    
+    const auto& gradient_wrt_output = this->get_prev_error_signals(tag);
+    auto& gradient_wrt_input = this->get_error_signals(tag);
+
+
+    int gradient_wrt_output_Height = gradient_wrt_output.Height();
+    int gradient_wrt_output_Width = gradient_wrt_output.Width();
+    for(int i=0; i<childs.size();i++)
+    {
+      auto* const gradient_wrt_input_cast = dynamic_cast<
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum.hpp:206]: What if this is not the type? Needs error message.

File: include/lbann/layers/transform/cross_grid_sum_slice.hpp [hunk 12]
```diff
@@ -163,6 +163,7 @@ class cross_grid_sum_slice_layer : public data_type_layer<TensorDataType>
 
   void bp_setup_gradient_wrt_inputs(El::Int mini_batch_size) override
   {
+    auto childs = this->get_child_layers();
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum_slice.hpp:166]:
```suggestion
    auto children = this->get_child_layers();
```

File: include/lbann/layers/transform/cross_grid_sum_slice.hpp [hunk 13]
```diff
@@ -204,7 +205,14 @@ class cross_grid_sum_slice_layer : public data_type_layer<TensorDataType>
     El::Transpose(temp_output, transposed_output);
     transposed_output.Resize(mloc * subgrid_comm_size, nloc);
 
-    gradient_wrt_input_cast->Resize(this->get_input_size(), mini_batch_size);
+    // gradient_wrt_input_cast->Resize(this->get_input_size(), mini_batch_size);
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum_slice.hpp:208]:
```suggestion
```

File: include/lbann/layers/transform/cross_grid_sum_slice.hpp [hunk 14]
```diff
@@ -204,7 +205,14 @@ class cross_grid_sum_slice_layer : public data_type_layer<TensorDataType>
...
+    for(int i=0; i<childs.size();i++)
+    {
+      auto* const gradient_wrt_input_cast_layer = dynamic_cast<
```
Comment (tbennun) [inline include/lbann/layers/transform/cross_grid_sum_slice.hpp:211]: needs error handling

File: include/lbann/layers/transform/split.hpp [hunk 15]
```diff
@@ -142,19 +142,19 @@ class split_layer : public data_type_layer<TensorDataType>
...
       for (int i = 0; i < this->get_num_children(); ++i) {
-        tag = childs[i]->get_parallel_strategy().sub_branch_tag;
+        tag = childs[i]->get_grid_tag();
 
         El::LockedView(this->get_activations(i),
                        this->get_branch_tag_input(tag - 1));
       }
     }
-    else {
+  else {
```
Comment (tbennun) [inline include/lbann/layers/transform/split.hpp:157]: please run `clang-format` on the code

File: include/lbann/layers/transform/sum.hpp [hunk 16]
```diff
@@ -136,19 +136,19 @@ class sum_layer : public data_type_layer<TensorDataType>
...
 #endif // LBANN_HAS_DISTCONV
     auto& output = this->get_activations();
-    auto parents = this->get_parent_layers();
+    const auto& parents = this->get_parent_layers();
 
-    if (this->is_subgraph_parallelism_enabled() &&
-        this->get_parallel_strategy().enable_subgraph == true) {
-      auto subgrid_tags = (*this->m_parent_tags);
+    if (this->subgraph_parallelism_execution()) {
+      // auto subgrid_tags = (*this->m_parent_tags);
```
Comment (tbennun) [inline include/lbann/layers/transform/sum.hpp:142]:
```suggestion
```

File: include/lbann/layers/transform/sum.hpp [hunk 17]
```diff
@@ -136,19 +136,19 @@ class sum_layer : public data_type_layer<TensorDataType>
...
       int tag = 0;
 
       std::vector<bool> is_initialized_tensor(this->m_num_spliting_groups,
                                               false);
 
       // Copy data internally with same branch tag
       for (int i = 0; i < this->get_num_parents(); ++i) {
-        tag = subgrid_tags[i];
+        // tag = subgrid_tags[i];
```
Comment (tbennun) [inline include/lbann/layers/transform/sum.hpp:150]:
```suggestion
```

File: include/lbann/layers/transform/sum.hpp [hunk 18]
```diff
@@ -248,11 +248,11 @@ class sum_layer : public data_type_layer<TensorDataType>
...
   {
     int tag = 0;
+    const auto& parents = this->get_parent_layers();
     const auto& gradient_wrt_output = this->get_prev_error_signals();
 
-    if (this->is_subgraph_parallelism_enabled() &&
-        this->get_parallel_strategy().enable_subgraph == true) {
-      auto subgrid_tags = (*this->m_parent_tags);
+    if (this->subgraph_parallelism_execution()) {
+      // auto subgrid_tags = (*this->m_parent_tags);
```
Comment (tbennun) [inline include/lbann/layers/transform/sum.hpp:255]:
```suggestion
```

File: src/layers/activations/softmax.cpp [hunk 19]
```diff
@@ -187,6 +187,7 @@ void softmax_layer<TensorDataType, Layout, Device>::fp_compute()
                       *this->m_workspace,
                       this->threshold_val,
                       this->m_mode);
+  std::cout<<"softmax Height:"<<this->get_prev_activations().Height()<<" Width:"<<this->get_prev_activations().Width()<<"\n";
```
Comment (tbennun) [inline src/layers/activations/softmax.cpp:190]:
```suggestion
```

File: src/layers/data_type_layer.cpp [hunk 20]
```diff
@@ -668,38 +668,39 @@ void data_type_layer<InputTensorDataType, OutputTensorDataType>::setup_matrices(
...
+  //               " layer \"",
+  //               this->get_name(),
+  //               "\" ",
+  //               "on invalid grid ",
+  //               "(grid tag ",
+  //               tag,
+  //               ", ",
+  //               grids.size(),
+  //               " grids available)");
+  // }
```
Comment (tbennun) [inline src/layers/data_type_layer.cpp:701]:
```suggestion
```

File: src/layers/data_type_layer.cpp [hunk 21]
```diff
@@ -668,38 +668,39 @@ void data_type_layer<InputTensorDataType, OutputTensorDataType>::setup_matrices(
...
-                "\" ",
-                "on invalid grid ",
-                "(grid tag ",
-                tag,
-                ", ",
-                grids.size(),
-                " grids available)");
-  }
-  this->set_grid_tag(tag);
+  // this->reset_mygrid(grids[tag]);
```
Comment (tbennun) [inline src/layers/data_type_layer.cpp:703]:
```suggestion
```

File: src/layers/data_type_layer.cpp [hunk 22]
```diff
@@ -838,20 +842,24 @@ void data_type_layer<InputTensorDataType, OutputTensorDataType>::setup_matrices(
       temp_grad = output_mat_builder->MakeEmpty(grid, 0);
     }
 
-    auto subgrid_tags = *(this->m_parent_tags);
+    // auto subgrid_tags = *(this->m_parent_tags);
```
Comment (tbennun) [inline src/layers/data_type_layer.cpp:845]:
```suggestion
```

File: src/layers/data_type_layer.cpp [hunk 23]
```diff
@@ -838,20 +842,24 @@ void data_type_layer<InputTensorDataType, OutputTensorDataType>::setup_matrices(
...
+      // for (int parent_index = 0; parent_index < int(parents.size());
+      //      ++parent_index) {
+      //   if (parents[parent_index]->get_grid_tag() - 1 == count) {
+      //     subgrid_tensor =
+      //       input_mat_builder->MakeEmpty(*grids[count],
+      //                                    0);
+      //     count++;
+      //     break;
+      //   }
+      // }
```
Comment (tbennun) [inline src/layers/data_type_layer.cpp:862]:
```suggestion
```

File: src/proto/factories/layer_graph_factory.cpp [hunk 24]
```diff
@@ -192,10 +192,10 @@ construct_layer_graph(lbann_comm* comm,
...
     ps.depth_groups = proto_layer.parallel_strategy().depth_groups();
     ps.depth_splits = proto_layer.parallel_strategy().depth_splits();
-    ps.enable_subgraph = proto_layer.parallel_strategy().enable_subgraph();
-    ps.sub_branch_tag = proto_layer.parallel_strategy().sub_branch_tag();
-    ps.sub_branch_resource_percentage =
-      proto_layer.parallel_strategy().sub_branch_resource_percentage();
+    // ps.enable_subgraph = proto_layer.parallel_strategy().enable_subgraph();
+    // ps.sub_branch_tag = proto_layer.parallel_strategy().sub_branch_tag();
+    // ps.sub_branch_resource_percentage =
+    //   proto_layer.parallel_strategy().sub_branch_resource_percentage();
```
Comment (tbennun) [inline src/proto/factories/layer_graph_factory.cpp:198]:
```suggestion
```

File: src/proto/layers.proto [hunk 25]
```diff
@@ -1552,11 +1552,11 @@ message ParallelStrategy {
   google.protobuf.Int64Value grid_tag = 21;
 
   /// @todo Remove
-  int64 sub_branch_tag = 15;
+  /// int64 sub_branch_tag = 15;
   /// @todo Remove
```
Comment (tbennun) [inline src/proto/layers.proto:1556]:
```suggestion
```

File: src/proto/layers.proto [hunk 26]
```diff
@@ -1552,11 +1552,11 @@ message ParallelStrategy {
   google.protobuf.Int64Value grid_tag = 21;
 
   /// @todo Remove
```
Comment (tbennun) [inline src/proto/layers.proto:1554]:
```suggestion
```

File: src/proto/layers.proto [hunk 27]
```diff
@@ -1552,11 +1552,11 @@ message ParallelStrategy {
...
-  int64 sub_branch_tag = 15;
+  /// int64 sub_branch_tag = 15;
   /// @todo Remove
-  int64 sub_branch_resource_percentage = 16;
+  /// int64 sub_branch_resource_percentage = 16;
   /// @todo Remove
```
Comment (tbennun) [inline src/proto/layers.proto:1558]:
```suggestion
```

File: src/proto/layers.proto [hunk 28]
```diff
@@ -1552,11 +1552,11 @@ message ParallelStrategy {
...
-  bool enable_subgraph = 17;
+  /// bool enable_subgraph = 17;
```
Comment (tbennun) [inline src/proto/layers.proto:1559]:
```suggestion
```

File: src/models/model.cpp [hunk 29]
```diff
@@ -723,8 +723,15 @@ void model::setup(size_t max_mini_batch_size,
 
   setup_layer_topology();
   setup_layer_execution_order();
+  setup_layer_grid_tags(grids_);
+  
   if (this->is_subgraph_parallelism_enabled()) {
-    setup_subgrids();
+  	for (auto& layer: this->get_layers())
```
Comment (tbennun) [inline src/models/model.cpp:729]: please run clang-format on your code. See [CONTRIBUTING.md](https://github.com/LLNL/lbann/blob/develop/CONTRIBUTING.md) for more information.

File: src/models/model.cpp [hunk 30]
```diff
@@ -884,79 +848,35 @@ void model::check_subgraph_parallelism()
...
   const auto& layers = this->get_layers();
   const El::Int num_layers = layers.size();
-  int myrank = El::mpi::Rank();
   for (El::Int node = 0; node < num_layers; ++node) {
-    auto layer_subgrid_ranks = layers[node]->get_subgrid_ranks();
-    if (layer_subgrid_ranks.find(myrank) != layer_subgrid_ranks.end() ||
-        layers[node]->get_name() == "layer1") {
-
-      layers[node]->set_run_layer_in_subgraph();
+    if(layers[node]->get_type() == "cross_grid_sum"){
```
Comment (tbennun) [inline src/models/model.cpp:854]: why are there special cases for layer types?

File: src/models/model.cpp [hunk 31]
```diff
@@ -884,79 +848,35 @@ void model::check_subgraph_parallelism()
...
+      layers[node]->set_subgraph_parallelism_execution();
```
Comment (tbennun) [inline src/models/model.cpp:855]: Could this not be in the constructor of the `cross_grid_sum` class?

File: python/lbann/models/__init__.py [hunk 32]
```diff
@@ -3,3 +3,6 @@
 from lbann.models.resnet import ResNet, ResNet18, ResNet34, ResNet50, ResNet101, ResNet152
 from lbann.models.roberta import RoBERTa
 from lbann.models.transformer import Transformer, TransformerEncoderLayer, TransformerDecoderLayer
+from lbann.models.subgraph.transformer import Transformer as TransformerSubGraph
```
Comment (tbennun) [inline python/lbann/models/__init__.py:6]: There is quite a bit of code duplication here, and the code is not up to date with the batched head implementation. What I'd suggest is that TransformerSubGraph would be a subclass of Transformer that sets up grid tags. We can discuss.

**Review by tbennun — CHANGES_REQUESTED** [review]
Overall the PR looks good, nice cleanup! There are a few minor comments in the code, one question about the behavior of certain layers, and one major comment in the Python frontend. We can discuss as necessary.

File: ci_test/unit_tests/test_unit_subgrid_split.py (hunk 4 above)
Comment (aj-prime) [reply ci_test/unit_tests/test_unit_subgrid_split.py:102]: Yes, we go over grid tags and check for subgraph parallelism

File: ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py (hunk 5 above)
Comment (aj-prime) [reply ci_test/unit_tests/test_unit_subgrid_cross_grid_sum_opt.py:1]: We can. Let's discuss it in the meeting.

File: include/lbann/layers/layer.hpp (hunk 8 above)
Comment (aj-prime) [reply include/lbann/layers/layer.hpp:877]: Model-level: Is subgraph parallelism enabled for Model?
Layer-level: Does this layer need subgraph execution (like split and sum layers)
Process-level: Will this layer run in the given process (For e.g. some layers will only on a subset of processes defined by grid tag)

File: include/lbann/layers/transform/cross_grid_sum.hpp (hunk 11 above)
Comment (aj-prime) [reply include/lbann/layers/transform/cross_grid_sum.hpp:206]: I can add it here but there are lot of places where we are just converting like this

File: include/lbann/layers/layer.hpp (hunk 8 above)
Comment (tbennun) [reply include/lbann/layers/layer.hpp:877]: 1. when you say "enabled for Model" do you mean "enabled for *this* Model?"
2. the name "process-level execution" for "is this executed on this process" may be unclear with respect to the others
3. Please add the explanations to the code itself as comments so it's clear for a reader and not just me

thanks!

File: include/lbann/layers/transform/cross_grid_sum.hpp (hunk 11 above)
Comment (tbennun) [reply include/lbann/layers/transform/cross_grid_sum.hpp:206]: OK, but I am not sure about the STAR/VC/ELEMENT type being always the same.

File: python/lbann/models/__init__.py (hunk 32 above)
Comment (tbennun) [reply python/lbann/models/__init__.py:6]: Post-discussion note: we will leave this for another PR.

**Comment by tbennun** [comment]
@aj-prime I'll approve the PR after dealing with the 5 unresolved comments

File: src/models/model.cpp (hunk 30 above)
Comment (aj-prime) [reply src/models/model.cpp:854]: Sub-graph parallelism is not a default case for these layers

**Comment by aj-prime** [comment]
Resolved all comments

**Review by tbennun — APPROVED** [review]
LGTM now, thank you!

**Review by benson31 — APPROVED** [review]

**Review by bvanessen — APPROVED** [review]

**Pull request merged**